packaging/_parser.py

```diff
@@ -68,8 +68,18 @@ def parse_named_requirement(requirement: str) -> Requirement:
         IDENTIFIER extras (URL_SPEC | specifier) (SEMICOLON marker_expr)? END
     """
     tokens = Tokenizer(requirement)
-    tokens.expect("IDENTIFIER", error_message="Expression must begin with package name")
-    name = tokens.read("IDENTIFIER").text
+    name_tokens = []
+    tokens.peek()
+    while tokens.next_token.name in NAME_SUITABLE_TOKENS:
```

Comment:
Looks like `tokens.next_token` has a type of `Optional[Token]`.

https://github.com/pypa/packaging/blob/606c71acce93d04e778cdbdea16231b36d6b870f/packaging/_tokenizer.py#L81

Line 75 that you added seems to also lean into that typing. Does the following make sense?

```suggestion
    while tokens.next_token and tokens.next_token.name in NAME_SUITABLE_TOKENS:
```